components/merkledb/examples/migration_with_iter_loop.rs

```diff
@@ -0,0 +1,162 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Shows how to migrate database data with the `MigrationHelper::iter_loop`. The migration follows the following scenario:
+//!
+//! 1. We create and fill database with random data according to schema defined in the
+//!   `migration::v1` module with the `create_initial_data` method.
+//! 2. We create `MigrationHelper` for this database.
+//! 3. We perform migration from the `v1` schema to the `v2` schema
+//!   with the help of the `create_migration` and `migrate_wallets` methods.
+//!   The method transforms the data in the old schema to conform to the new schema.
+//!   Wallets migration is performed with `MigrationHelper::iter_loop`.
+//!   The old data is **not** removed at this stage; rather, it exists alongside
+//!   the migrated data. This is useful in case the migration needs to be reverted for some reason.
+//! 4. We complete the migration by calling `flush_migration`. This moves the migrated data
+//!   to its intended place and removes the old data marked for removal.
+
+use exonum_merkledb::{
+    access::Prefixed,
+    migration::{flush_migration, MigrationHelper},
+    ObjectHash, Result as DbResult,
+};
+
+use migration::{
+    check_data_after_flush, check_data_after_merge, check_data_before_flush, create_initial_data,
+    v1, v2,
+};
+
+mod migration;
+
+fn create_migration(helper: &mut MigrationHelper) -> DbResult<()> {
+    println!("\nStarted migration");
+
+    {
+        let new_data = helper.new_data();
+        let old_data = helper.old_data();
+
+        let old_schema = v1::Schema::new(old_data);
+        let mut new_schema = v2::Schema::new(new_data);
+
+        // Move `ticker` and `divisibility` to `config`.
+        let config = v2::Config {
+            ticker: old_schema.ticker.get().unwrap(),
+            divisibility: old_schema.divisibility.get().unwrap_or(0),
+        };
+        new_schema.config.set(config);
+        // Mark these two indexes for removal.
+        new_data.create_tombstone("ticker");
+        new_data.create_tombstone("divisibility");
+    }
+
+    // Migrate wallets using `MigrationHelper::iter_loop`:
+    // `Wallet::public_key` field will be removed.
+    // `Wallet::history_hash` field will be added.
+    // Wallets and history from username Eve will be removed.
+    migrate_wallets(helper)
+}
+
+/// Provides migration of wallets with `MigrationHelper::iter_loop`.
+/// `iter_loop` is designed to allow to merge changes to the database from time to time,
+/// so we are migrating wallets in chunks here.
+fn migrate_wallets(helper: &mut MigrationHelper) -> DbResult<()> {
+    helper.iter_loop(|helper, iters| {
+        let old_schema = v1::Schema::new(helper.old_data());
+        let mut new_schema = v2::Schema::new(helper.new_data());
+
+        const CHUNK_SIZE: usize = 1_000;
```

Comment:
Nit: I'd mention here that `CHUNK_SIZE` is chosen such that we can safely store the part of migration in RAM.